lib/sentry/transport.ex

```diff
@@ -29,18 +29,39 @@ defmodule Sentry.Transport do
       case Envelope.to_binary(envelope) do
         {:ok, body} ->
           {endpoint, headers} = get_endpoint_and_headers()
-          post_envelope_with_retries(client, endpoint, headers, body, retries)
+          post_envelope_with_retries(client, endpoint, headers, body, retries, envelope.items)
 
         {:error, reason} ->
           {:error, ClientError.new({:invalid_json, reason})}
       end
 
     _ = maybe_log_send_result(result, envelope.items)
-
     result
   end
 
-  defp post_envelope_with_retries(client, endpoint, headers, payload, retries_left) do
+  def record_discarded_event(reason, event_item) do
+    _ =
+      if is_list(event_item) do
+        Enum.map(event_item, fn event ->
+          ClientReport.add_discarded_event({reason, Envelope.get_data_category(event)})
+        end)
+      else
+        ClientReport.add_discarded_event({reason, Envelope.get_data_category(event_item)})
+      end
```

Comment:
1. `List.wrap/1`
2. `Enum.map/2` wastes memory here, use `Enum.each/2` since we ignore the result.

```suggestion
    _ =
      event_item
      |> List.wrap()
      |> Enum.each(&ClientReport.add_discarded_event({reason, Envelope.get_data_category(&1)}))
```